### ISSUE TITLE
Redirection for already established TCP connection after Podman container live migration 

### ISSUE THREAD

**Issue by alidhamieh** [checkpoint-restore/criu#1598]
**Description**

I did Podman container live migration to another machine using Podman checkpoint/restore with "--tcp-established" argument. My question: is there a way to redirect an opened TCP connection to the new destination machine? I tested iptables forwarding and ssh tunneling but established tcp client hanged when i installed the forwarding rules. I want to have seamless live migration so already established tcp connection continue on the new server. The client connects to a gateway linux and this gw redirects traffic to the intended server. And when live migration happens, the gateway will redirect the client traffic to the new server location. The client uses the gateway IP as destination IP. This works for new tcp connections but not for already established tcp connections. Usually for Nat or similar there are rules which only react on new connections with syn set. Maybe we need different rules for existing connections (unset syn for established state)?

To make a Linux machine acts as Gateway, i used iptables with the following rules:
echo 1 | sudo tee /proc/sys/net/ipv4/ip_forward
iptables -A FORWARD -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT

iptables -t nat -A PREROUTING -p tcp --dport 80 -j DNAT --to-destination x.x.x.x:80
iptables -t nat -A POSTROUTING -j MASQUERADE)

x.x.x.x is the destination of the migration ip

**Steps to reproduce the issue:**
1. Send GET request to a webserver for a big file size
2. In the middle of the transfer, do live migration using commit_cmd = 'sudo podman container checkpoint %s --export=/tmp/%s.tar.gz --tcp-established' % (container, container, ) 
3. scp /tmp/%s.tar.gz  to the destination of migration
4.  And restore: restore_cmd = "sudo podman container restore --import=%s%s.tar.gz --tcp-established" % (remote_path, container, )
5. Run the above rules to make the middle man acts as a gateway

**Describe the results you received:**

I ran tcpdump; it seems there is no redirection happened for already established connection. Redirection only happened for new connections.
Redirection worked for new tcp connections but not for already established tcp connections.
The clients hanged when they are on established tcp state after checkpoint/restore steps.

**Describe the results you expected:**

Redirection works for established tcp connections. Clients continue the tcp connection with the new migrated server.

**Comment by adrianreber** [comment]
Can you add examples of IP addresses of the client, the container, the gateway before and after the migration.

**Comment by alidhamieh** [comment]
> Can you add examples of IP addresses of the client, the container, the gateway before and after the migration.

The client was on ip1. The source container was on the same machine as the gateway and has internal IP of 10.88.0.67. The destination of the container was on a different machine than the client and the gateway and it has an internal IP of 10.88.0.67 (same as the source container). Does this answer the question?

**Comment by adrianreber** [comment]
What IP does the gateway have? So the client with *ip1* talks to the gateway with which IP?

**Comment by alidhamieh** [comment]
The source container and the gateway are on the same physical machine (different than the client's physical machine). The gateway has public IP 35.x.x.x different than the client ip1 and the source container internal IP 10.88.0.67. The source container, gateway and the destination container are on Google Cloud. The client is on my local machine.

**Comment by alidhamieh** [comment]
> What IP does the gateway have? So the client with _ip1_ talks to the gateway with which IP?

The source container and the gateway are on the same physical machine (different than the client's physical machine). The gateway has public IP 35.x.x.x different than the client ip1 and the source container internal IP 10.88.0.67. The source container, gateway and the destination container are on Google Cloud. The client is on my local machine.

**Comment by adrianreber** [comment]
Okay. Does the gateway IP change during the migration? I understand it in such a way that you migrate a container from host1 to to host2. And the gateway is on host1 and host2.

I am sorry, but I still do not understand your setup. Can you give example IP addresses of all involved parts. Gateway and client. What does change?

**Comment by alidhamieh** [comment]
> Okay. Does the gateway IP change during the migration? I understand it in such a way that you migrate a container from host1 to to host2. And the gateway is on host1 and host2.
No the gateway ip does not change during migration. No the gateway is on host1 only (not on host2).
> I am sorry, but I still do not understand your setup. Can you give example IP addresses of all involved parts. Gateway and client. What does change?
[
![setup](https://user-images.githubusercontent.com/66278376/131514933-b7763121-3a58-45f3-9b71-f29120107d5d.png)
](url)

**Comment by adrianreber** [comment]
You migrate the container to `Destination of migration` and client still talks to the gateway, right?

Can you share a tcpdump just before the migration and after the migration, so that we can see how your network traffic is going to the `Destination of migration`.

**Comment by alidhamieh** [comment]
> You migrate the container to `Destination of migration` and client still talks to the gateway, right?
> 
> Can you share a tcpdump just before the migration and after the migration, so that we can see how your network traffic is going to the `Destination of migration`.

Yes, I migrate the container to Destination of migration and client still talks to the gateway. I did tcpdump for that before but did not save it. There are two cases: (1) client connects after migration but to the gateway (not to the destination of migration) and it was able to receive the data from the gateway while the gateway was getting the data from the destination of migration. The packet source ip and destination ip from the client was changed at the gateway to the gateway ip and the destination of migration, respectively. so the client communicates to the gateway and the gateway communicates with the destination of migration and the gateway sends back the data to the client using its ip address (received from the destination of migration). (2) While client is connected (in established connection state) to the gateway, the migration happened. Before migration the client was receiving data and sends acknowledge to the source container. After migration, the client stopped receiving any data (no data was received) and the destination of migration did not receive any packets from the gateway and the gateway did not receive any packets from the destination of migration.
Note: I will do the test again and save the tcpdump output and shared here.
